src/memory/memory.c

```diff
@@ -150,13 +150,30 @@ static void _clean_chunk(uint8_t** chunk_bytes)
     uint8_t* __attribute__((__cleanup__(_clean_chunk))) __attribute__((unused)) var##_bytes = \
         (var).bytes;
 
+#ifndef TESTING
+static uint32_t _address_offset(memory_area_t area)
+{
+    switch (area) {
+    case MEMORY_AREA_SHARED_DATA:
+        return FLASH_SHARED_DATA_START;
+    case MEMORY_AREA_APP_DATA:
+        return FLASH_APPDATA_START;
+    case MEMORY_AREA_SMARTEEPROM:
+        return FLASH_SMARTEEPROM_START;
+    default:
+        Abort("unknown memory area");
+    }
+}
+#endif
+
 // chunk must have size CHUNK_SIZE. if chunk is NULL, the chunk is erased,
 // i.e. filled with 0xFF.
-static bool _write_to_address(uint32_t addr, uint8_t* chunk)
+static bool _write_to_address(memory_area_t area, uint32_t addr, uint8_t* chunk)
 {
 #ifdef TESTING
-    return memory_write_to_address_mock(addr, chunk);
+    return memory_write_to_address_mock(area, addr, chunk);
```

Comment:
this function can then simply switch on the area to write to different static vars in RAM instead of trying to infer it from the actual address.